### ISSUE TITLE
[FEAT] Interaction term between two comparisons

### ISSUE THREAD

**Issue by V-Lamp** [moj-analytical-services/splink#2413]
### Is your proposal related to a problem?
Terms with independent comparisons, e.g. Postcode and City can be very correlated, so two independent comparisons for postcode & city will lead to overestimating match score when both match, or underestimating it when only one matches.

### Describe the solution you'd like
Some mechanism to score the interaction between two comparisons (usually negatively, like in term frequency).

### Describe alternatives you've considered
So far I have put city as a lower comparison level to postcode, but I expect this problem of correlated comparisons to be more general.

### Additional context
Creating an interaction term is a common mechanism in dealing with correlation in Machine Learning, e.g.
if x1 and x2 are correlated, you can add a term x1*x2 in your model, e.g. `y = a*x1+b*x2+c*x1*x2+d`

**Comment by zmbc** [comment]
@RobinL's solution is equivalent to method 1 in S4 of [the appendix of the fastLink paper](https://imai.fas.harvard.edu/research/files/linkage-app.pdf), and works great for some use-cases, but doesn't allow e.g. specifying c1 * c2 and c2 * c3 without c1 * c2 * c3. For that, the second method in that appendix based on log-linear models is the solution, and there is a discussion of adding it to Splink here: https://github.com/moj-analytical-services/splink/discussions/1310

**Comment by V-Lamp** [comment]
Thank you for the response around using AND, I think I can incorporate it. However, one complexity that usually comes in practice is that comparisons usually have more than one comparison level.

For example: 
```python
postcode_comparison = [exact_match, area_match, sector_match]
address_comparison = [exact_match, street_name_match, jaro_winkler_match]
location_comparison = ???
```

I would need to make 9 levels (3 * 3) with AND, plus the other 3 + 3 levels, to define a location_comparison (so 3*3 + 3 + 3 levels in total). The challenging thing then is to find what is the right order for these 15 comparison levels, since ordering has a very significant effect. In my case, I actually have more that 3 levels, more like 6-8.

Have you found yourself in this combinatorial explosion and then further ordering problem?

**Comment by RobinL** [comment]
Yeah - you're right to highlight these challenges.  It's typically best to try and order in terms of 'better matches higher' - start with the most precise matches and work your way down.  Although I appreciate it's not always obvious in practice; you may need some trial and error.

I agree that the combinatorial explosion problem is real, but on a large dataset having (say) 9 comparison levels is totally fine.  Ultimately, each one corresponds to two parameters to estimate, so 18 parameters is not very many at all (compared to, say, other ML approaches which can have thousands).

In our production models we tend to have between about 2 and 10 comparison levels per comparison
